Source/Core/Resource.js

```diff
@@ -0,0 +1,534 @@
+                    uri.query = undefined;
+                }
+
+                // Remove the fragment as it's not sent with a request
+                uri.fragment = undefined;
+
+                this._url = uri.toString();
+
+                if (this._isDirectory) {
+                    this._url = appendForwardSlash(this._url);
+                }
+            }
+        },
+
+        /**
+         * The file extension of the resource.
+         *
+         * @memberof Resource.prototype
+         * @type {String}
+         *
+         * @readonly
+         */
+        extension: {
+            get: function() {
+                return getExtensionFromUri(this._url);
+            }
+        },
+
+        /**
+         * True if the Resource refers to a directory. A trailing forward slash will be maintained.
+         *
+         * @memberof Resource.prototype
+         * @type {Boolean}
+         */
+        isDirectory: {
+            get: function() {
+                return this._isDirectory;
+            },
+            set: function(value) {
+                if (this._isDirectory !== value) {
+                    var url = this._url;
+                    if (value) {
+                        this._url = appendForwardSlash(url);
+                    } else if (url[url.length-1] === '/') {
+                        this._url = url.slice(0, -1);
+                    }
+
+                    this._isDirectory = value;
+                }
+            }
+        },
+
+        /**
+         * True if the Resource refers to a data URI.
+         *
+         * @memberof Resource.prototype
+         * @type {Boolean}
+         */
+        isDataUri: {
+            get: function() {
+                return isDataUri(this._url);
+            }
+        }
+    });
+
+    /**
+     * Returns the url, optional with the query string and processed by a proxy.
+     *
+     * @param {Boolean} [query=false] If true, the query string is included.
+     * @param {Boolean{ [proxy=false] If true, the url is processed the proxy object if defined.
+     */
+    Resource.prototype.getUrlComponent = function(query, proxy) {
+        var uri = new Uri(this._url);
+
+        if (query) {
+            uri.query = stringifyQuery(this._queryParameters);
+        }
+
+        // objectToQuery escapes the placeholders.  Undo that.
+        var url = uri.toString().replace(/%7B/g, '{').replace(/%7D/g, '}');
+
+        var template = this._templateValues;
+        var keys = Object.keys(template);
+        if (keys.length > 0) {
+            for (var i = 0; i < keys.length; i++) {
+                var key = keys[i];
+                var value = template[key];
+                url = url.replace(new RegExp('{' + key + '}', 'g'), value);
+            }
+        }
+        if (proxy && defined(this.proxy)) {
+            url = this.proxy.getURL(url);
+        }
+        return url;
+    };
+
+    /**
+     * Adds query parameters
```

Comment:
Provide better doc here and in `addTemplateValues`.  Users can also just modify `queryParemeters` directly, right?  So perhaps provide an explanation of when they would use this instead?